proposals/NNNN-dxil18.md

```diff
@@ -0,0 +1,282 @@
+```
+
+The `kDxilNodeIOFlagsTag` is based on the `DXIL::NodeIOFlags` and
+`DXIL::NodeIOKind` enumerations as described below:
+
+```c++
+enum class NodeIOFlags : uint32_t {
+    None = 0x0,
+    Input = 0x1,
+    Output = 0x2,
+    ReadWrite = 0x4,
+    EmptyRecord = 0x8, // EmptyNodeOutput[Array], EmptyNodeInput
+    NodeArray = 0x10,  // NodeOutputArray, EmptyNodeOutputArray
+
+    // Record granularity (enum in 2 bits)
+    ThreadRecord = 0x20,   // [RW]ThreadNodeInputRecord, ThreadNodeOutputRecords
+    GroupRecord = 0x40,    // [RW]GroupNodeInputRecord, GroupNodeOutputRecords
+    DispatchRecord = 0x60, // [RW]DispatchNodeInputRecord
+    RecordGranularityMask = 0x60,
+
+    NodeIOKindMask = 0x7F,
+
+    TrackRWInputSharing = 0x100, // TrackRWInputSharing tracked on all non-empty
+                                 // input/output record/node types
+
+    // Mask for node/record properties beyond NodeIOKind
+    RecordFlagsMask = 0x100,
+    NodeFlagsMask = 0x100,
+  };
+
+  enum class NodeIOKind : uint32_t {
+    Invalid = 0,
+
+    EmptyInput =
+        (uint32_t)NodeIOFlags::EmptyRecord | (uint32_t)NodeIOFlags::Input,
+    NodeOutput =
+        (uint32_t)NodeIOFlags::ReadWrite | (uint32_t)NodeIOFlags::Output,
+    NodeOutputArray = (uint32_t)NodeIOFlags::ReadWrite |
+                      (uint32_t)NodeIOFlags::Output |
+                      (uint32_t)NodeIOFlags::NodeArray,
+    EmptyOutput =
+        (uint32_t)NodeIOFlags::EmptyRecord | (uint32_t)NodeIOFlags::Output,
+    EmptyOutputArray = (uint32_t)NodeIOFlags::EmptyRecord |
+                       (uint32_t)NodeIOFlags::Output |
+                       (uint32_t)NodeIOFlags::NodeArray,
+
+    DispatchNodeInputRecord =
+        (uint32_t)NodeIOFlags::Input | (uint32_t)NodeIOFlags::DispatchRecord,
+    GroupNodeInputRecords =
+        (uint32_t)NodeIOFlags::Input | (uint32_t)NodeIOFlags::GroupRecord,
+    ThreadNodeInputRecord =
+        (uint32_t)NodeIOFlags::Input | (uint32_t)NodeIOFlags::ThreadRecord,
+
+    RWDispatchNodeInputRecord = (uint32_t)NodeIOFlags::ReadWrite |
+                                (uint32_t)NodeIOFlags::Input |
+                                (uint32_t)NodeIOFlags::DispatchRecord,
+    RWGroupNodeInputRecords = (uint32_t)NodeIOFlags::ReadWrite |
+                              (uint32_t)NodeIOFlags::Input |
+                              (uint32_t)NodeIOFlags::GroupRecord,
+    RWThreadNodeInputRecord = (uint32_t)NodeIOFlags::ReadWrite |
+                              (uint32_t)NodeIOFlags::Input |
+                              (uint32_t)NodeIOFlags::ThreadRecord,
+
+    GroupNodeOutputRecords = (uint32_t)NodeIOFlags::ReadWrite |
+                             (uint32_t)NodeIOFlags::Output |
+                             (uint32_t)NodeIOFlags::GroupRecord,
+    ThreadNodeOutputRecords = (uint32_t)NodeIOFlags::ReadWrite |
+                              (uint32_t)NodeIOFlags::Output |
+                              (uint32_t)NodeIOFlags::ThreadRecord,
+  };

```

Comment:
I think this might be more easily consumable as a table either with a single field that lists the flags enabled or maybe even with a column for each flag.